### PR TITLE
feat(cache): consolidate caching API and strategies

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by hypeJunction** [comment]
I can't see any more Elgg code. This is good for review, need to figure what's up with tests. I am leaving it all for a while, I need a break

File: engine/classes/Elgg/Database/EntityTable.php [hunk 1]
```diff
@@ -627,4 +626,110 @@ public function disableEntities($owner_guid) {
...
+		_elgg_invalidate_cache_for_entity($guid);
+
+		$qb = Delete::fromTable('entities');
+		$qb->where($qb->compare('guid', '=', $guid, ELGG_VALUE_INTEGER));
+
+		$deleted = $this->db->deleteData($qb);
+
+		_elgg_services()->hooks->getEvents()->triggerAfter('delete', $entity->type, $entity);
+
+		return (bool) $deleted;
```
Comment (mrclay) [inline engine/classes/Elgg/Database/EntityTable.php]: When would we want to return false?

File: engine/classes/Elgg/Cache/MetadataCache.php [hunk 2]
```diff
@@ -227,39 +254,16 @@ public function populateFromEntities(...$guids) {
...
-			}
-
-			if (isset($this->ids[$guid][$name])) {
-				$this->ids[$guid][$name] = (array) $this->ids[$guid][$name];
-				$this->ids[$guid][$name][] = $id;
-			} else {
-				$this->ids[$guid][$name] = $id;
-			}
+			$row->value = ($row->value_type === 'text') ? $row->value : (int) $row->value;
+			$metadata = new ElggMetadata($row);
```
Comment (mrclay) [inline engine/classes/Elgg/Cache/MetadataCache.php]: What do we get for the overhead of creating these objects?

File: engine/classes/Elgg/Database/PrivateSettingsTable.php [hunk 3]
```diff
@@ -132,29 +134,40 @@ public function getAll($entity_guid) {
...
-			(entity_guid, name, value) VALUES
-			(:entity_guid, :name, :value)
-			ON DUPLICATE KEY UPDATE value = :value
-		";
-		$params = [
-			':entity_guid' => (int) $entity_guid,
-			':name' => (string) $name,
-			':value' => (string) $value,
-		];
+		_elgg_invalidate_cache_for_entity($entity_guid);
```
Comment (mrclay) [inline engine/classes/Elgg/Database/PrivateSettingsTable.php]: This stays?

Comment (mrclay) [reply engine/classes/Elgg/Database/PrivateSettingsTable.php]: That's fine

**Review by mrclay — COMMENTED** [review]
Bad idea for me to try reviewing this on my tiny phone!

File: engine/classes/Elgg/Cache/MetadataCache.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Cache/MetadataCache.php]: Not sure how we would measure it. Definitely better than iterating through bunch of array and querying the DB to compare metadata names and ids

File: engine/classes/Elgg/Database/EntityTable.php (hunk 1 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Database/EntityTable.php]: I am not sure. Just moved this from ElggEntity

File: engine/classes/Elgg/Cache/MetadataCache.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Cache/MetadataCache.php]: We can also just cache attributes

File: engine/classes/Elgg/Cache/MetadataCache.php [hunk 4]
```diff
@@ -270,25 +274,25 @@ public function populateFromEntities(...$guids) {
...
 			'limit' => 0,
-			'callback' => false,
+			'callback' => function($e) {
+				return (int) $e->entity_guid;
+			},
 			'selects' => ['SUM(LENGTH(n_table.value)) AS bytes'],
 			'order_by' => 'n_table.entity_guid, n_table.time_created ASC',
 			'group_by' => 'n_table.entity_guid',
+			'having' => [
+				"bytes > $limit OR bytes < 0",
```
Comment (hypeJunction) [inline engine/classes/Elgg/Cache/MetadataCache.php]: Should be the opposite. Need to filter those that have heavy meta

File: engine/classes/Elgg/Cache/SessionCache.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+	/**
+	 * Create a new cache under a namespace
+	 *
+	 * @param string $namespace Namespace
+	 *
+	 * @return CompositeCache
+	 */
+	protected function create($namespace) {
+		return new CompositeCache($namespace, $this->config, ELGG_CACHE_RUNTIME);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Cache/SessionCache.php]: I am contemplating making it persistent

File: engine/classes/Elgg/Database/EntityTable.php [hunk 6]
```diff
@@ -627,4 +626,110 @@ public function disableEntities($owner_guid) {
...
+			access_show_hidden_entities($entity_disable_override);
+			elgg_set_ignore_access($ia);
+		}
+
+		$entity_disable_override = access_get_show_hidden_status();
+		access_show_hidden_entities(true);
+		$ia = elgg_set_ignore_access(true);
+
+		// Now delete the entity itself
+		$entity->deleteMetadata();
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: Ned to add foreign key and get rid of this madness

File: engine/classes/Elgg/Database/PrivateSettingsTable.php [hunk 7]
```diff
@@ -99,21 +96,24 @@ public function get($entity_guid, $name) {
...
 
-		$query = "
-			SELECT * FROM {$this->table}
-			WHERE entity_guid = :entity_guid
-		";
-		$params = [
-			':entity_guid' => (int) $entity_guid,
-		];
+		$values = $this->cache->load($entity_guid);
+		if ($values) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/PrivateSettingsTable.php]: Eh?

File: engine/classes/Elgg/Database/PrivateSettingsTable.php [hunk 8]
```diff
@@ -132,29 +134,40 @@ public function getAll($entity_guid) {
...
-		];
+		_elgg_invalidate_cache_for_entity($entity_guid);
+
+		$qb = Select::fromTable('private_settings');
+		$qb->select('id')
+			->where($qb->compare('name', '=', $name, ELGG_VALUE_STRING))
+			->andWhere($qb->compare('entity_guid', '=', $entity_guid, ELGG_VALUE_INTEGER));
 
-		$result = $this->db->insertData($query, $params);
+		$row = $this->db->getDataRow($qb);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/PrivateSettingsTable.php]: Not really efficient. Better strategy would be to cache settings with as database rows and grab IDs from there

File: engine/classes/ElggUpgrade.php [hunk 9]
```diff
@@ -15,6 +15,11 @@
  *
  * @package Elgg.Admin
  * @access private
+ *
+ * @property int $is_complete
```
Comment (hypeJunction) [inline engine/classes/ElggUpgrade.php]: Completed

**Comment by hypeJunction** [comment]
I give up. All tests are passing locally, but not on Travis. Faster tests come with all that shitty global state that spills over.

**Comment by hypeJunction** [comment]
Tests are green now

File: 1 [hunk 10]
```diff
@@ -0,0 +1,21 @@
+pick 2e5fe21b8 feat(cache): consolidate caching API and strategies
```
Comment (jdalsem) [inline 1]: where is this file coming from

Comment (hypeJunction) [reply 1]: Not sure. Something happened during rebase. Will remove

File: engine/classes/Elgg/Database/EntityTable.php [hunk 11]
```diff
@@ -455,9 +464,16 @@ public function fetch(QueryBuilder $query, array $options = []) {
 			return [];
 		}
 
-		$preload = array_filter($results, function($e) {
+		$preload = array_filter($results, function ($e) {
 			return $e instanceof ElggEntity;
 		});
+		/* @var $preload ElggEntity[] */
+
+		if ((bool) $options['limit'] === false || $options['limit'] > elgg_get_config('default_limit')) {
```
[the remark this answers is not in the document]
Comment (hypeJunction) [reply engine/classes/Elgg/Database/EntityTable.php]: Allows allow plugins

File: engine/classes/ElggPlugin.php [hunk 12]
```diff
@@ -1155,8 +1149,9 @@ private function setStatus($active) {
 			$result = remove_entity_relationship($this->guid, 'active_plugin', $site->guid);
 		}
 
+		$this->invalidateCache();
+		_elgg_services()->plugins->invalidateCache($this->getID());
```
Comment (hypeJunction) [inline engine/classes/ElggPlugin.php]: Move to entity method

**Review by hypeJunction — COMMENTED** [review]
Need to bring back entity cache. Locking session cache creates a situation where caches are not invalidated when entity is updated with ignored access

**Comment by hypeJunction** [comment]
I think I got it all sorted out now.

**Comment by mrclay** [comment]
Works fine for me.

**Comment by jeabakker** [comment]
This PR accidentally (I assume) undid #11503

**Pull request merged**